enhancements/ingress/lb-subnet-selection-aws.md

```diff
@@ -0,0 +1,597 @@
+after upgrading to 4.y doesn't require a cluster admin to delete the LoadBalancer-type
+service:
+
+1. Cluster admin confirms that initially `service.beta.kubernetes.io/aws-load-balancer-subnets`
+   is set on the LoadBalancer-type service managed by the IngressController.
+2. The cluster admin upgrades OpenShift to v4.y and the service annotation is not
+   changed or removed.
+3. After upgrading, the IngressController will emit a `LoadBalancerProgressing` condition
+   with `Status: True` because the spec's `Subnets` (an empty slice `[]`) does not equal
+   the current annotation.
+4. Instead of recreating the LoadBalancer-type service, the cluster admin will set
+   `spec.endpointPublishingStrategy.loadBalancer.providerParameters.aws.subnets` to the
+   current value of the `service.beta.kubernetes.io/aws-load-balancer-subnets` service.
+5. The Ingress Operator will resolve the `LoadBalancerProgressing` condition back to
+   `Status: False` as long as `Subnets` is equivalent to `service.beta.kubernetes.io/aws-load-balancer-subnets`.
+
+### API Extensions
+
+This proposal will modify the `IngressController` API by adding a new
+field called `Subnets` of type `[]string` to the `AWSLoadBalancerParameters`
+struct type.
+
+TODO: What will install-time option modify?
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+This enhancement is directly enabling the https://issues.redhat.com/browse/XCMSTRAT-545
+feature for ROSA Hosted Control Planes.
+
+#### Standalone Clusters
+
+This proposal does not have any special implications for standalone clusters.
+
+#### Single-node Deployments or MicroShift
+
+This proposal does not have any special implications for single-node
+deployments or MicroShift.
+
+### Implementation Details/Notes/Constraints
+
+When an IngressController is created with
+`spec.endpointPublishingStrategy.loadBalancer.providerParameters.aws.subnets`
+specified, the Ingress Operator will set the
+`service.beta.kubernetes.io/aws-load-balancer-subnets` annotation
+on the LoadBalancer-type Service.
+
+However, if `spec.endpointPublishingStrategy.loadBalancer.providerParameters.aws.subnets`
+is updated on an existing IngressController, the Ingress Operator will NOT
+update `service.beta.kubernetes.io/aws-load-balancer-subnets` until the
+cluster admin deletes the LoadBalancer-type service as indicated by the
+`LoadBalancerProgressing` status (see [Effectuating Subnet Updates](#effectuating-subnet-updates)).
+When the Ingress Operator recreates the LoadBalancer-type service, it will
+then configure `service.beta.kubernetes.io/aws-load-balancer-subnets` with
+the new `Subnets` value.
+
+The `status.endpointPublishingStrategy.loadBalancer.providerParameters.aws.subnets`
+will always reflect the _actual_ subnet value by being mirroring the value of
+`service.beta.kubernetes.io/aws-load-balancer-subnets` on the service. We
+anticipate that cluster admins may have added an unmanaged
```

Comment:
I wouldn't mention the unsupported option unless it adds meaning here.  I think this paragraph is supposed to point out that the `spec` and `status` may not look synchronized with the unsupported service annotation.  If so, it could use a little wordsmithing.

Reply:
> I wouldn't mention the unsupported option unless it adds meaning here.

This was added per Miciah's request https://github.com/openshift/enhancements/pull/1595#discussion_r1552520873.

> I think this paragraph is supposed to point out that the spec and status may not look synchronized with the unsupported service annotation. If so, it could use a little wordsmithing.

The purpose of this paragraph is to say "the status should mirror the annotation, not the spec". I attempted to rewrite into a list to make it clearer.

Reply:
The EP absolutely should call out considerations such as a cluster-admin's use of an unsupported option given that the platform didn't prevent the cluster-admin from using it.  Some customers assume that if OpenShift lets them do it, then it's supported.  Ideally, we should block the use of unsupported configuration or block upgrades of clusters with unsupported configuration, but failing that, we should make a reasonable effort not to break clusters that have unsupported configuration.